conda_smithy/schema.py

```diff
@@ -0,0 +1,1165 @@
+    automatic version updates/migrations for feedstocks.
+    """
+
+    automerge: Optional[Union[bool, BotConfigAutoMergeChoice]] = Field(
+        False,
+        description="Automatically merge PRs if possible",
+    )
+
+    check_solvable: Optional[bool] = Field(
+        None,
+        description="Open PRs only if resulting environment is solvable.",
+        exclude=True,  # Will not be rendered in the model dump
+    )
+
+    inspection: Optional[Union[bool, BotConfigInspectionChoice]] = Field(
+        None,
+        description="Method for generating hints or updating recipe",
+        exclude=True,  # Will not be rendered in the model dump
+    )
+
+    abi_migration_branches: Optional[List[Union[str, int, float]]] = Field(
+        None,
+        description="List of branches for additional bot migration PRs",
+        exclude=True,  # Will not be rendered in the model dump
+    )
+
+    version_updates_random_fraction_to_keep: Optional[float] = Field(
+        None,
+        description="Fraction of versions to keep for frequently updated packages",
+        exclude=True,  # Will not be rendered in the model dump
+    )
+
+
+class CondaBuildConfig(BaseModel):
+    pkg_format: Optional[Literal["tar", 1, 2, "1", "2"]] = Field(
+        description="The package version format for conda build.",
+        default=2,
+    )
+
+    zstd_compression_level: Optional[int] = Field(
+        default=16,
+        description="""The compression level for the zstd compression algorithm for
+            .conda artifacts. conda-forge uses a default value of 16 for a good
+            compromise of performance and compression.""",
+    )
+
+    error_overlinking: Optional[bool] = Field(
+        default=False,
+        description="""
+            Enable error when shared libraries from transitive dependencies are
+            directly  linked  to any executables or shared libraries in  built
+            packages. For more details, see the
+            [conda build documentation](https://docs.conda.io/projects/conda-build/en/stable/resources/commands/conda-build.html).
+            """,
+    )
+
+
+class CondaForgeDocker(BaseModel):
+    executable: Optional[str] = Field(
+        description="The executable for Docker", default="docker"
+    )
+
+    fallback_image: Optional[str] = Field(
+        description="The fallback image for Docker",
+        default="quay.io/condaforge/linux-anvil-comp7",
+    )
+
+    command: Optional[str] = Field(
+        description="The command to run in Docker", default="bash"
+    )
+
+    interactive: Optional[Union[bool, Nullable]] = Field(
+        description="Whether to run Docker in interactive mode",
+        default=None,
+        exclude=True,  # Will not be rendered in the model dump
+    )
+
+    #########################################
+    #### Deprecated Docker configuration ####
+    #########################################
+    image: Optional[Union[str, Nullable]] = Field(
```

Comment:
How does the deprecation hint validation work here? I don't see any metadata here that annotates this as deprecated? Or is the list maintained elsewhere?

Reply:
Ah, I saw in `build_with_mambabuild` that we have a `deprecated=True` kwarg in Field? That seems to be pydantic v1 only, and apparently we need `json_schema_extra={'deprecated': True}` now.

Reply:
More info https://github.com/pydantic/pydantic/issues/2255

Reply:
Marked with `deprecated=True` kwarg in Field. We have our own custom validator for deprecated fields.

Reply:
Will this interfere with the JSON schema output? Maybe it's worth either switching to `json_schema_extra={'deprecated': True}` as the marker or having the schema dumper do that from `deprecated=True`, since pydantic v2 won't do that for us. This way the IDE tools can use the [`deprecated`](https://json-schema.org/draft/2020-12/json-schema-validation#name-deprecated) marker to style keys and throw warnings. That said, it's not yet ready on VS Code, so feel free to ignore. I'm just concern we forget about this courtesy detail in the future :)

Reply:
No, the schema dumper sets `deprecated=True` correctly.
For eg:
```json
"compiler_stack": {
      "anyOf": [
        {
          "type": "string"
        },
        {
          "type": "null"
        }
      ],
      "default": "comp7",
      "deprecated": true,
      "description": "\n        Compiler stack environment variable. This is used to specify the compiler\n        stack to use for builds. Deprecated.\n\n        .. code-block:: yaml\n\n            compiler_stack: comp7\n        ",
      "title": "Compiler Stack"
    },
```